mlir/lib/Quantum/Transforms/AdjointPatterns.cpp

```diff
@@ -0,0 +1,224 @@
+// limitations under the License.
+
+#define DEBUG_TYPE "adjoint"
+
+#include <algorithm>
+#include <iterator>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Errc.h"
+
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+#include "Quantum/IR/QuantumOps.h"
+#include "Quantum/Transforms/Patterns.h"
+
+using namespace llvm;
+using namespace mlir;
+using namespace catalyst::quantum;
+
+namespace {
+
+template <class T> T isInstanceOf(Operation &op)
+{
+    if (op.getName().getStringRef() == T::getOperationName())
+        return cast<T>(op);
+    else
+        return nullptr;
+}
+
+/// Copy the region of the adjoint operation `op` to the POI specified by the `rewriter`. Build and
+/// return the value mapping `bvm`.
+Value copyAdjointVerbatim(AdjointOp op, PatternRewriter &rewriter, IRMapping &bvm)
+{
+    Block &b = op.getRegion().front();
+    for (auto i = b.begin(); i != b.end(); i++) {
+        if (YieldOp yield = isInstanceOf<YieldOp>(*i)) {
+            assert(++i == b.end() &&
+                   "quantum.yield must be the last operation of an adjoint block");
+            return bvm.lookupOrDefault(yield->getOperand(0));
+        }
+        else {
+            rewriter.insert(i->clone(bvm));
+        }
+    }
+    assert(false && "quantum.yield must present in the adjoint region");
+}
+
+struct AdjointSingleOpRewritePattern : public mlir::OpRewritePattern<AdjointOp> {
+    using mlir::OpRewritePattern<AdjointOp>::OpRewritePattern;
+
+    /// In essence, we build a map from values mentiond in the source data flow to the values of the
+    /// program where quantum control flow is reversed. Most of the time, there is a 1-to-1
+    /// correspondence with a notable exception caused by `insert`/`extract` API asymetry.
+    mlir::LogicalResult matchAndRewrite(AdjointOp adjoint,
+                                        mlir::PatternRewriter &rewriter) const override
+    {
+        LLVM_DEBUG(dbgs() << "Adjointing the following:\n" << adjoint << "\n");
+        Location loc = adjoint.getLoc();
+        MLIRContext *ctx = adjoint.getContext();
+        assert(adjoint.getRegion().hasOneBlock());
+
+        // First, copy the classical computations directly to the target POI and build the classical
+        // value mapping dictionary.
+        auto classicalMapping = ({
+            IRMapping out;
+            Block &b = adjoint.getRegion().front();
+            for (auto i = b.begin(); i != b.end(); i++) {
+                if (i->getName().getStringRef().find("quantum") != std::string::npos) {
+                    /* Skip quantum operations */
+                }
+                else {
+                    LLVM_DEBUG(dbgs() << "classical operation: " << *i << "\n");
+                    rewriter.insert(i->clone(out));
+                }
+            }
+            out;
```

Comment:
What is this line doing?

Reply:
That is the syntax that I was mentioning in my comment above. I think @grwlf, it might be best to change it. Here's the link that discusses this [syntax](https://stackoverflow.com/a/6305429)

> It assigns user input value to a and prints it out. it is done by using a Statement Expression.
>
> Statement Expressions are a [gnu gcc compiler extension](http://gcc.gnu.org/onlinedocs/gcc/Statement-Exprs.html#Statement-Exprs) and are not supported by the C/C++ standards. Hence, any code which uses statement expression is not standard conforming and non-portable.

Reply:
Ah thanks, agreed!

Reply:
In general, I do realize that this syntax is unusual, but IMO cleaning function-level variable scope from temporary variables is a very important benefit improving code safety and readability.

Reply:
Done. Thanks again, I was sure it is a standard!